README

LIBSYSCONFCPUS 0.5
==================

by Kevin Pulo, kev@pulo.com.au
http://www.kev.pulo.com.au/libsysconfcpus/
Copyright (c) 2008-2009
Licensed under the GNU GPL v2 or higher, as per the COPYING file.


DESCRIPTION
-----------

libsysconfcpus is an $LD_PRELOAD library that intercepts calls to sysconf(),
and adjusts the responses for _SC_NPROCESSORS_CONF and _SC_NPROCESSORS_ONLN,
which are used to determine the number of processors available (either
configured or online) at runtime in a system.  Unfortunately, some closed-
source software assumes that all of these processors may be used for
computation, which is an assumption that is frequently not true on multiuser
and HPC systems.  Thus, libsysconfcpus provides a simple, non-intrusive way of
adjusting the behaviour of such software (when it is dynamically linked).

In addition, libsysconfcpus is a small, simple example of how to use
$LD_PRELOAD to do function call interception on glibc and POSIX systems.  It is
hoped that it is useful not only inherently, but also for learning about
$LD_PRELOAD, and as a template or skeleton LD_PRELOAD library.  Its
from-scratch development was used to illustrate LD_PRELOAD usage in the "Fun
with LD_PRELOAD" presentation by the author at linux.conf.au 2009 in Hobart.


INSTALLATION
------------

Standard autoconf process:  ./configure && make && make install


USAGE
-----

    sysconfcpus [options] <program> [args...]

Valid options are:

    -v, --version   Display the sysconfcpus version
    -h, --help      Display this help
    -d, --debug     Display debug info

    -n, --num <n>   Number of processors (both conf and onln)
    -c, --conf <n>  Number of processors (conf(igured))
    -o, --onln <n>  Number of processors (online)


EXAMPLEs
--------

$ test-sysconfcpus 
sysconf(_SC_NPROCESSORS_CONF) = 2
sysconf(_SC_NPROCESSORS_ONLN) = 2
$ sysconfcpus -n 8 test-sysconfcpus 
sysconf(_SC_NPROCESSORS_CONF) = 8
sysconf(_SC_NPROCESSORS_ONLN) = 8
$ sysconfcpus -c 8 -o 6 test-sysconfcpus 
sysconf(_SC_NPROCESSORS_CONF) = 8
sysconf(_SC_NPROCESSORS_ONLN) = 6
$ 


LIMITATIONS
-----------

- Only dynamically linked binaries


FEEDBACK
--------

Comments, feature suggestions, bug reports, patches, etc are most
welcome, please send them to Kevin Pulo <kev@pulo.com.au>.


SEE ALSO
--------

- ltrace
- strace




ChangeLog

v0.5: Extra features
- Use $LIBSYSCONFCPUS_CONF and $LIBSYSCONFCPUS_ONLN, with $LIBSYSCONFCPUS as
  a fallback.
- The sysconfcpus and sysconfcpus-debug shell scripts accept basic command
  line parameters.
- Add simple test-sysconfcpus program.


v0.4: Adjust sysconf() behaviour
- For sysconf(_SC_NPROCESSORS_CONF) and sysconf(_SC_NPROCESSORS_ONLN), if
  $LIBSYSCONFCPUS is set to a number, use that as the override return value.


v0.3: Intercept sysconf()
- Add sysconf() function that calls the real sysconf().


v0.2: Library debug and init
- Add dprintf(), controlled by $LIBSYSCONFCPUS_DEBUG.
- Add init/fini library startup/shutdown functions.
- Add sysconfcpus and sysconfcpus-debug helper shell scripts.


v0.1: Empty base project
- Library does nothing.




src/test-sysconfcpus.c
/* libsysconfcpus: adjust number of CPUs reported by sysconf()
 * Copyright (C) 2008-2009 Kevin Pulo
 * 
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 * 
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 * 
 * Contact:
 * Kevin Pulo
 * kev@pulo.com.au
 */

#include <stdio.h>
#include <unistd.h>

int main(int argc, char *argv[]) {
	long num_configured;
	long num_online;

	num_configured = sysconf(_SC_NPROCESSORS_CONF);
	printf("sysconf(_SC_NPROCESSORS_CONF) = %ld\n", num_configured);

	num_online = sysconf(_SC_NPROCESSORS_ONLN);
	printf("sysconf(_SC_NPROCESSORS_ONLN) = %ld\n", num_online);

	return 0;
}



src/libsysconfcpus.c
/* libsysconfcpus: adjust number of CPUs reported by sysconf()
 * Copyright (C) 2008-2009 Kevin Pulo
 * 
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 * 
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 * 
 * Contact:
 * Kevin Pulo
 * kev@pulo.com.au
 */

#include <stdio.h>
#include <stdarg.h>
#include <stdlib.h>

/* Using _GNU_SOURCE as the manpage suggests can be problematic.
 * We only want the GNU extensions for dlfcn.h, not others like stdio.h.
 * But dlfcn.h checks __USE_GNU, not _GNU_SOURCE.  __USE_GNU is set by
 * features.h when _GNU_SOURCE is set, but features.h has already been loaded
 * by stdio.h, so it doesn't get loaded again.  Ugh.  Setting _GNU_SOURCE
 * earlier drags in other junk (eg. dprintf).  Simplest is just to use
 * __USE_GNU directly.
 */
#define __USE_GNU
#include <dlfcn.h>
#undef __USE_GNU

#include <unistd.h>



static int debug = 0;

static const char libsysconfcpus_debug_envvar[] = "LIBSYSCONFCPUS_DEBUG";

static int do_debug() {
	return (debug || getenv(libsysconfcpus_debug_envvar));
}

static void debugprintf(char *fmt, ...) {
	va_list ap;
	if (do_debug()) {
		va_start(ap, fmt);
		vfprintf(stderr, fmt, ap);
		va_end(ap);
		fflush(stderr);
	}
}





static void _libsysconfcpus_init(void) __attribute__((constructor));
static void _libsysconfcpus_fini(void) __attribute__((destructor));

static void _libsysconfcpus_init(void) {
	debugprintf("libsysconfcpus: starting up\n");
}

static void _libsysconfcpus_fini(void) {
	debugprintf("libsysconfcpus: shutting down\n");
}



static const char libsysconfcpus_envvar[] = "LIBSYSCONFCPUS";
static const char libsysconfcpus_conf_envvar[] = "LIBSYSCONFCPUS_CONF";
static const char libsysconfcpus_onln_envvar[] = "LIBSYSCONFCPUS_ONLN";

long sysconf(int name) {
	static void *libc_handle = NULL;
	static long (*underlying)(int name);
	long retval;
	long v;
	const char *env;
	const char *envvar;
	const char *envvar_use;
	const char *err;
	char *endptr;

	debugprintf("libsysconfcpus: sysconf(%d): entered\n", name);

	if (!libc_handle) {
#if defined(RTLD_NEXT)
		libc_handle = RTLD_NEXT;
#else
		libc_handle = dlopen("libc.so.6", RTLD_LAZY);
#endif
		debugprintf("libsysconfcpus: sysconf: libc_handle = 0x%x\n", libc_handle);
		if (!libc_handle) {
			fprintf(stderr, "libsysconfcpus: Error: Unable to find libc.so: %s\n", dlerror());
			exit(1);
		}

		dlerror();
		underlying = dlsym(libc_handle, "sysconf");
		debugprintf("libsysconfcpus: sysconf: underlying = 0x%x\n", underlying);
		err = dlerror();
		if (err) {
			debugprintf("libsysconfcpus: sysconf: err = \"%s\"\n", err);
		}
		if (!underlying || err) {
			fprintf(stderr, "libsysconfcpus: Error: Unable to find the underlying sysconf(): %s\n", dlerror());
			exit(1);
		}
	}

	debugprintf("libsysconfcpus: about to call underlying sysconf()\n");
	retval = (*underlying)(name);
	debugprintf("libsysconfcpus: underlying sysconf(%d) = %ld\n", name, retval);

	envvar = NULL;
	switch (name) {
		case _SC_NPROCESSORS_CONF:
			debugprintf("libsysconfcpus: sysconf(_SC_NPROCESSORS_CONF)\n");
			envvar = libsysconfcpus_conf_envvar;
			break;
		case _SC_NPROCESSORS_ONLN:
			debugprintf("libsysconfcpus: sysconf(_SC_NPROCESSORS_ONLN)\n");
			envvar = libsysconfcpus_onln_envvar;
			break;
	}
	if (envvar) {
		envvar_use = envvar;
		env = getenv(envvar_use);
		if (!env) {
			envvar_use = libsysconfcpus_envvar;
			env = getenv(envvar_use);
		}
		if (env) {
			debugprintf("libsysconfcpus: $%s = \"%s\"\n", envvar_use, env);
			v = strtol(env, &endptr, 10);
			if (endptr == env) {
				debugprintf("libsysconfcpus: Warning: $%s does not contain a number, cannot override\n", envvar_use);
			} else {
				retval = v;
			}
		} else {
			debugprintf("libsysconfcpus: Warning: neither $%s nor $%s are set, cannot override\n", envvar, libsysconfcpus_envvar);
		}
	}

	debugprintf("libsysconfcpus: sysconf(%d) = %ld\n", name, retval);

	return retval;
}
